### PR TITLE
simplifier pass: cherry pick fix for template_04

### PR DESCRIPTION
**Pull request by gxyd**
Cherry-pick commit for fix for `integration_tests/template_04.f90` from https://github.com/lfortran/lfortran/pull/3748

### CODE REVIEW

File: src/libasr/codegen/llvm_utils.cpp [hunk 1]
```diff
@@ -2098,6 +2098,10 @@ namespace LCompilers {
                 if( ASR::is_a<ASR::Character_t>(*alloc_type->m_type) ) {
                     lfortran_str_copy(dest, src, true, *module, *builder, context);
                 } else {
+                    if( ASRUtils::is_array(alloc_type->m_type) ) {
+                        llvm::Type* loadedType = src->getType()->getPointerElementType();
```
Comment (gxyd) [inline src/libasr/codegen/llvm_utils.cpp:2102]: @Thirumalai-Shaktivel , can you help me with the correct change for LLVM 17 here? The CI currently passes for all versions of LLVM, except LLVM 17. With LLVM 17, I get the below error when building it:
```console
FAILED: src/libasr/CMakeFiles/asr.dir/codegen/llvm_utils.cpp.o 
sccache /usr/bin/c++ -DHAVE_TARGET_X86=1 -DHAVE_WHEREAMI=1 -DHAVE_ZLIB=1 -DLCOMPILERS_FAST_ALLOC=1 -I/home/runner/work/lfortran/lfortran/src/libasr/.. -isystem /home/runner/micromamba/envs/lf/include -Wall -Wextra -g -ggdb -std=gnu++17 -fPIC -fno-rtti -Wno-deprecated-declarations -MD -MT src/libasr/CMakeFiles/asr.dir/codegen/llvm_utils.cpp.o -MF src/libasr/CMakeFiles/asr.dir/codegen/llvm_utils.cpp.o.d -o src/libasr/CMakeFiles/asr.dir/codegen/llvm_utils.cpp.o -c /home/runner/work/lfortran/lfortran/src/libasr/codegen/llvm_utils.cpp
/home/runner/work/lfortran/lfortran/src/libasr/codegen/llvm_utils.cpp: In member function ‘void LCompilers::LLVMUtils::deepcopy(llvm::Value*, llvm::Value*, LCompilers::ASR::ttype_t*, llvm::Module*, std::map<std::__cxx11::basic_string<char>, std::map<std::__cxx11::basic_string<char>, int> >&)’:
/home/runner/work/lfortran/lfortran/src/libasr/codegen/llvm_utils.cpp:[210](https://github.com/lfortran/lfortran/actions/runs/11164726735/job/31034757411?pr=4966#step:5:211)2:66: error: ‘class llvm::Type’ has no member named ‘getPointerElementType’; did you mean ‘getStructElementType’?
 2102 |                         llvm::Type* loadedType = src->getType()->getPointerElementType();
      |                                                                  ^~~~~~~~~~~~~~~~~~~~~
      |                                                                  getStructElementType
[74/118] Building CXX object src/lfortran/CMakeFiles/lfortran_lib.dir/parser/parser.cpp.o
```

File: integration_tests/CMakeLists.txt [hunk 2]
```diff
@@ -1429,14 +1429,14 @@ RUN(NAME template_array_03 LABELS llvm llvm_wasm llvm_wasm_emcc llvmStackArray w
...
 RUN(NAME template_matrix_01 LABELS llvm llvm_wasm llvm_wasm_emcc llvm17)
-# RUN(NAME template_matrix_test LABELS llvm llvm_wasm llvm_wasm_emcc llvm17 EXTRAFILES
-#     template_semigroup.f90
-#     template_semiring.f90
-#     template_monoid.f90
-#     template_unitring.f90
-#     template_field.f90
-#     template_matrix.f90
-#     )
+RUN(NAME template_matrix_test LABELS llvm llvm_wasm llvm_wasm_emcc llvm17 EXTRAFILES
```
Comment (Pranavchiku) [inline integration_tests/CMakeLists.txt:1432]:
```suggestion
RUN(NAME template_matrix_test LABELS llvm llvm_wasm llvm_wasm_emcc EXTRAFILES
```

Comment (Pranavchiku) [reply integration_tests/CMakeLists.txt:1432]: I would do this and merge to unblock ourselves on it.

**Comment by certik** [comment]
Here is the error:
```
/home/runner/work/lfortran/lfortran/src/libasr/codegen/llvm_utils.cpp: In member function ‘void LCompilers::LLVMUtils::deepcopy(llvm::Value*, llvm::Value*, LCompilers::ASR::ttype_t*, llvm::Module*, std::map<std::__cxx11::basic_string<char>, std::map<std::__cxx11::basic_string<char>, int> >&)’:
/home/runner/work/lfortran/lfortran/src/libasr/codegen/llvm_utils.cpp:2102:66: error: ‘class llvm::Type’ has no member named ‘getPointerElementType’; did you mean ‘getStructElementType’?
 2102 |                         llvm::Type* loadedType = src->getType()->getPointerElementType();
      |                                                                  ^~~~~~~~~~~~~~~~~~~~~
      |                                                                  getStructElementType
```
You can't query an LLVM pointer of its type anymore. Instead, you have to use the mapping we have in the LLVM backend to determine the type for a pointer, see other code there. Alternatively, you use the new "load" mechanism that does it automatically.

**Comment by gxyd** [comment]
There isn't much to review here, as the code was written by Gagan, I made some changes for LLVM 17 (atleast for building), I'll merge this PR now.

**Pull request merged**